core/src/main/java/com/linecorp/armeria/client/Endpoint.java

```diff
@@ -83,36 +93,71 @@ public static Endpoint of(String host) {
 
     /**
      * Creates a new host {@link Endpoint}.
+     *
+     * @deprecated Use {@link #of(String, int)} and {@link #withWeight(int)},
+     *             e.g. {@code Endpoint.of("foo.com", 80).withWeight(500)}.
      */
+    @Deprecated
     public static Endpoint of(String host, int port, int weight) {
+        return of(host, port).withWeight(weight);
+    }
+
+    private static Endpoint create(String host, int port) {
         requireNonNull(host, "host");
-        validatePort("port", port);
-        validateWeight(weight);
 
-        return new Endpoint(host, port, weight);
+        if (NetUtil.isValidIpV4Address(host)) {
+            return new Endpoint(host, host, port, DEFAULT_WEIGHT, true, false);
+        }
+
+        if (NetUtil.isValidIpV6Address(host)) {
+            final String ipV6Addr;
+            if (host.charAt(0) == '[') {
+                // Strip surrounding '[' and ']'.
+                ipV6Addr = host.substring(1, host.length() - 1);
+            } else {
+                ipV6Addr = host;
+            }
+            return new Endpoint(ipV6Addr, ipV6Addr, port, DEFAULT_WEIGHT, true, true);
+        }
+
+        return new Endpoint(host, null, port, DEFAULT_WEIGHT, false, false);
     }
 
     @Nullable
     private final String groupName;
     @Nullable
     private final String host;
+    @Nullable
+    private final String ipAddr;
     private final int port;
     private final int weight;
+    private final boolean hostIsIpAddr;
```

Comment:
You can also consider a `@Nullable` enum so it's just one field.

Reply:
Good idea. Now using a nullable enum.